### PR TITLE
Add VLAN tests

### PR DESCRIPTION
**Pull request by StepanVovkPLV**
This PR is grouped from smaller PR's

Adds following tests: 
```
- test_vlan_basic_functionality
- test_vlan_changing_default_pvid
- test_vlan_default_configuration_with_["broadcast", "multicast", "unicast", "unknown_unicast"]
- test_vlan_switching_vlan_modes_via_cli
- test_vlan_tagged_ports_with_["broadcast", "multicast", "unicast"]
```

### CODE REVIEW

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+        streams.update({f'{tg_ports[3]} -> {tg_ports[0]}': {
+            "type": "raw",
+            "protocol": "802.1Q",
+            "ip_source": dev_groups[tg_ports[3]][0]["name"],
+            "ip_destination": dev_groups[tg_ports[0]][0]["name"],
+            "src_mac": dst_mac,
+            "dst_mac": src_mac,
+            "vlanID": 1
+        }})
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py]: Convert it to a function that configures streams based on traffic type. There is much code duplication here.

**Review by yuriilisovskyi — CHANGES_REQUESTED** [review]

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        "src_mac": src_mac,
+        "dst_mac": dst_mac[traffic_type]
+    }})
+    if traffic_type == "multicast":
+        for stream in streams:
+            streams[stream].update({"rate": 50})
+    return streams
+
+
+@pytest.mark.parametrize("traffic_type", ["broadcast", "broadcast", "unicast", "unknown_unicast"])
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py]: why broadcast twice?

**Review by SerhiyBoikoPLV — APPROVED** [review]

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+              pytest.mark.usefixtures("cleanup_bridges", "cleanup_tgen")]
+
+
+async def test_vlan_basic_functionality(testbed):
+    """
+    Test Name: VLAN basic functionality
+    Test Suite: suite_functional_vlan
+    Test Overview: Test VLAN basic functionality with broadcast packet forwarding
+    Test Procedure:
+    1. Retrieve tgen and dent device.
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py]: This is not a test step. This is a step from your code only.
You need to separate the test steps from the implementation steps.

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_changing_default_pvid.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            {"port": 3, "settings": [{"vlan": 4094, "untagged": True, "pvid": True}]})
+
+
+async def test_vlan_changing_default_pvid(testbed):
+    """
+    Test Name: Changing default VLAN pvid
+    Test Suite: suite_functional_vlan
+    Test Overview: Test broadcast packet forwarding with non-default pvid
+    Test Procedure:
+    1. Retrieve tgen and dent device
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_changing_default_pvid.py]: Please use "test steps" and not code steps everywhere.

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_changing_default_pvid.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        "ip_source": tx_ports,
+        "ip_destination": rx_ports,
+        "src_mac": "02:00:00:00:00:01",
+        "dst_mac": "ff:ff:ff:ff:ff:ff",
+        "vlanID": vlan
+    } for vlan in packet_vids if vlan != "X"}
+
+    streams.update({"Untagged traffic": {
+        "type": "raw",
+        "protocol": "802.1Q",
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_changing_default_pvid.py]: Untagged but protocol 802.1q?

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            "ip_source": dev_groups[tg_ports[0]][0]["name"],
+            "ip_destination": [dev_groups[tg_ports[3]][0]["name"]],
+            "src_mac": src_mac,
+            "dst_mac": dst_mac[traffic_type],
+            "vlanID": vlan
+        } for vlan in packet_vids if vlan != "X"}
+
+        streams.update({f"{tg_ports[0]} -> X": {
+            "type": "raw",
+            "protocol": "802.1Q",
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py]: Re-check that you use the correct protocol in ALL untagged traffic items.

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+    # 5. Setup packet stream(s) for the packet:
+    streams = await set_up_traffic_streams(traffic_type, tg_ports, dev_groups)
+
+    await tgen_utils_setup_streams(tgen_dev, config_file_name=None, streams=streams)
+
+    # 6. Send traffic to rx_ports
+    # Sending traffic to learn MAC addresses
+    if traffic_type == "unicast":
+        await tgen_utils_start_traffic(tgen_dev)
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py:155]: Do you send bi-dir traffic? How this step can help to learn MACs?

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_switching_modes.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+              pytest.mark.usefixtures("cleanup_bridges", "cleanup_tgen")]
+
+
+async def test_vlan_switching_vlan_modes_via_cli(testbed):
+    """
+    Test Name: Switching port vlan modes via cli
+    Test Suite: suite_functional_vlan
+    Test Overview: Test broadcast packet forwarding after changing vlan port modes via cli
+    Test Procedure:
+    # 1. Initiate test params per first configuration of VLAN modes.
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_switching_modes.py]: Please remove hashes before the step number.

**Review by anton7811 — CHANGES_REQUESTED** [review]

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py (hunk 7 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py:155]: Not bi-directional traffic
1. Sending traffic 1st time so MAC's will learn on ports
2. Sending traffic and checking the losses

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
+import asyncio
+import pytest
+
+from dent_os_testbed.utils.test_utils.tgen_utils import (tgen_utils_get_dent_devices_with_tgen,
+                                                         tgen_utils_setup_streams, tgen_utils_start_traffic,
+                                                         tgen_utils_stop_traffic, tgen_utils_get_traffic_stats,
+                                                         tgen_utils_get_loss, tgen_utils_dev_groups_from_config,
+                                                         tgen_utils_traffic_generator_connect)
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py]: Little bit hard to read. Can you move each import to new line? Like:
`from dent_os_testbed.utils.test_utils.tgen_utils import (                                                      
      tgen_utils_get_dent_devices_with_tgen,                                                                        
    ....                                                                                                        
)`

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+async def test_vlan_basic_functionality(testbed):
+    """
+    Test Name: VLAN basic functionality
+    Test Suite: suite_functional_vlan
+    Test Overview: Test VLAN basic functionality with broadcast packet forwarding
+    Test Procedure:
+    1. Configure bridge; enslave ports to bridge. Set bridge, ports to `up` state
+    2. Set links to vlans per port_map configuration (all ports untagged, mixed vlans)
+    3. Map tx_ports and rx_ports
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py]: What does it mean?

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+
+async def test_vlan_basic_functionality(testbed):
+    """
+    Test Name: VLAN basic functionality
+    Test Suite: suite_functional_vlan
+    Test Overview: Test VLAN basic functionality with broadcast packet forwarding
+    Test Procedure:
+    1. Configure bridge; enslave ports to bridge. Set bridge, ports to `up` state
+    2. Set links to vlans per port_map configuration (all ports untagged, mixed vlans)
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py]: Since your last step verifies traffic on specific ports, be more specific in this step too. Specify which configuration is applied per ports

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_changing_default_pvid.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+
+async def test_vlan_changing_default_pvid(testbed):
+    """
+    Test Name: Changing default VLAN pvid
+    Test Suite: suite_functional_vlan
+    Test Overview: Test broadcast packet forwarding with non-default pvid
+    Test Procedure:
+    1. Configure bridge; enslave ports to bridge. Set bridge, ports to `up` state
+    2. Set links to vlans per port_map configuration (all ports untagged, same vlans)
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_changing_default_pvid.py]: Better to rephrase it to "Add ports to VLANs..."

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_changing_default_pvid.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        "type": "raw",
+        "protocol": "802.1Q",
+        "ip_source": tx_ports,
+        "ip_destination": rx_ports,
+        "src_mac": "02:00:00:00:00:01",
+        "dst_mac": "ff:ff:ff:ff:ff:ff",
+        "vlanID": vlan
+    } for vlan in packet_vids if vlan != "X"}
+
+    streams.update({"Untagged traffic": {
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_changing_default_pvid.py]: No need to use update() here, you can just add a value to the dictionary

**Review by yuriilisovskyi — CHANGES_REQUESTED** [review]

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py (hunk 10 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py]: Selecting from which port to which will be sending traffic

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+            "ip_destination": [dev_groups[tg_ports[3]][0]["name"]],
+            "src_mac": src_mac,
+            "dst_mac": dst_mac[traffic_type]
+        }
+        streams[f'{tg_ports[3]} -> {tg_ports[0]}'] = {
+            "type": "raw",
+            "protocol": "802.1Q",
+            "ip_source": dev_groups[tg_ports[3]][0]["name"],
+            "ip_destination": dev_groups[tg_ports[0]][0]["name"],
+            "src_mac": dst_mac,
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py]: Is this correct?

**Review by yuriilisovskyi — CHANGES_REQUESTED** [review]

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py (hunk 14 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py]: Yes, in case of unicast traffic we creating `traffic` from both ends end send them for 1 sec for sole purpose of learning macs on bridge. After generating traffic and checking losses

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py (hunk 10 above)
Comment (anton7811) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py]: This is what we discussed in other PR: implementation steps vs test steps.
Port mapping is implementation and has nothing to the test steps themselves. Step 3 could be removed and in steps 5 and 6 you can list exact port numbers (as you already partially did in step 6)

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_changing_default_pvid.py [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+async def test_vlan_changing_default_pvid(testbed):
+    """
+    Test Name: Changing default VLAN pvid
+    Test Suite: suite_functional_vlan
+    Test Overview: Test broadcast packet forwarding with non-default pvid
+    Test Procedure:
+    1. Configure bridge; enslave ports to bridge. Set bridge, ports to `up` state
+    2. Add ports to VLANs: all ports untagged, same vlans
+        Ports 1, 2, 3, 4 : vlan 4094, untagged, pvid
+    3. Map tx_ports and rx_ports
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_changing_default_pvid.py]: The same comment as for the test above.

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_changing_default_pvid.py [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+    await configure_bridge_setup(device, dut_ports, default_pvid=non_default_pvid)
+
+    # 2. Set links to vlans per port_map configuration (all ports untagged, same vlan)
+    port_map = ({"port": 0, "settings": [{"vlan": 4094, "untagged": True, "pvid": True}]},
+                {"port": 1, "settings": [{"vlan": 4094, "untagged": True, "pvid": True}]},
+                {"port": 2, "settings": [{"vlan": 4094, "untagged": True, "pvid": True}]},
+                {"port": 3, "settings": [{"vlan": 4094, "untagged": True, "pvid": True}]})
+    await configure_vlan_setup(device, port_map, dut_ports)
+
+    # 3. Map tx_ports and rx_ports
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_changing_default_pvid.py]: As I understand, real mapping happens on the line 99 in function get_traffic_port_vlan_mapping

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py (hunk 14 above)
Comment (anton7811) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py]: dst_mac is dict. Is this correct?

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+            "dst_mac": dst_mac[traffic_type],
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py]: Flip macs. So in dst you have src as well.

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py [hunk 18]
```diff
@@ -0,0 +1,96 @@
+import asyncio
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py]: The main concern about all these tests - there are too many test modules. Please group tests. IMO you can make 2-3 modules here.

**Review by anton7811 — CHANGES_REQUESTED** [review]

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py (hunk 18 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_basic_functionality.py]: Made it 3 files. Grouped tests related to default setups

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py [hunk 19]
```diff
@@ -0,0 +1,330 @@
+import asyncio
+import pytest
+
+from dent_os_testbed.lib.ip.ip_link import IpLink
+from dent_os_testbed.utils.test_utils.tgen_utils import (
+    tgen_utils_get_dent_devices_with_tgen,
+    tgen_utils_setup_streams, tgen_utils_start_traffic,
+    tgen_utils_stop_traffic, tgen_utils_get_traffic_stats,
+    tgen_utils_get_loss, tgen_utils_dev_groups_from_config,
+    tgen_utils_traffic_generator_connect)
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py]: Move each function name into separate line

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py [hunk 20]
```diff
@@ -0,0 +1,330 @@
...
+from dent_os_testbed.utils.test_utils.br_utils import (
+    configure_bridge_setup, configure_vlan_setup,
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py]: Same

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_switching_modes.py [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+    7. Set links to new vlans per second configuration of vlan modes
+        Port 1: vlan 5, untagged, pvid
+                vlan 101, tagged,
+                vlan 4094, untagged,
+        Port 2: vlan 5, untagged, pvid
+        Port 3: vlan 101, untagged, pvid
+        Port 4: vlan 4094, untagged, pvid
+    8. Setup broadcast packet stream(s) with vlans 0, 5, 101, 500, 4094 and untagged stream
+    9. Send traffic.
+    10. Verify traffic.
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_switching_modes.py]: Be more specific how do we verify traffic

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_tagged_ports.py [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+                vlan 24, untagged, pvid
+        Port 3: vlan 22
+                vlan 23,
+                vlan 24, untagged, pvid
+        Port 4:vlan 22
+                vlan 23,
+                vlan 24, untagged, pvid
+    3. Setup broadcast packet stream(s) with vlans 0, 1, 22, 23, 24 and untagged stream
+    4. Send traffic
+    5. Verify traffic
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_tagged_ports.py]: Same here

**Review by yuriilisovskyi — CHANGES_REQUESTED** [review]

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py (hunk 7 above)
Comment (anton7811) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py:155]: Do you expect that stream "->X" will be dropped? You do not add a stream to learn MAC for it. So it is not a symmetric case.

Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py:155]: Yes, `stream "->X"` will be dropped first time, second time when sending traffic we expect that no losses occur with this stream

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py [hunk 23]
```diff
@@ -0,0 +1,335 @@
...
+              pytest.mark.usefixtures("cleanup_bridges", "cleanup_tgen")]
+
+packet_vids = ['X', 1, 2]  # VLAN tag number contained in the transmitted packet
+port_map = ({"port": 0, "settings": [{"vlan": 1, "untagged": True, "pvid": True}]},
+            {"port": 1, "settings": [{"vlan": 1, "untagged": True, "pvid": True}]},
+            {"port": 2, "settings": [{"vlan": 1, "untagged": True, "pvid": True}]},
+            {"port": 3, "settings": [{"vlan": 1, "untagged": True, "pvid": True}]})
+
+
+async def set_up_traffic_streams(traffic_type, tg_ports, dev_groups):
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_default_configuration.py]: No need for async function here

**Review by yuriilisovskyi — APPROVED** [review]
